### DISCUSSION TITLE
zsh and iTerm2

### DISCUSSION THREAD

**Discussion by Tissot11** [ohmyzsh/ohmyzsh#10383] [Help] [answered]
I'm using iTerm2 and zsh on my Mac. I have a strange problem that whenever I try to reuse  a successfully executed long command after modifying it, the terminal prompt shows the correct command but on execution it fails because editing the command kept some of the the letter from previous command. These letters were invisible on the execution and only showed up after the failure of the new command. I paste below my .zshrc content. Can anyone suggest what is wrong and how to fix it. I had tried fixing before the escape character problem in iTerm2

```
#Command Prompt adjustments

PS1="[%{%F{blue}%}%n%{%f%}@%m-|%F{blue}%w-%F{blue}%*% %F{blue}%f|-%F{magenta}%~%}%{%F{blue}%}]-%F{magneta}%f%{%G🖖%} %f%}%{%G>%}%f "

PS1="%F{blue}[%}%n%{%f%}@%F{blue}%D{%a %b %e|%X%z(%Z)|}%F{blue}%f%F{magenta}%~%}%{%F{blue}%}]-%F{magneta}%f%{%G🖖%} %f%}%{%G>%}%f "

#Login to .bash_profile
alias profile='nano ~/.zshrc'

# To refresh the bash shell profile
alias reload='source ~/.zshrc'

#eval "$(ssh-agent -s)"
#ssh-add ~/.ssh/id_rsa

# To fix the mess of Sierra
ssh-add -K 2>/dev/null;

alias mv='mv -i'
alias cp='cp -i'
#alias ls='ls -lah'
alias ls='ls -lh'
alias df='df -H'
alias home='cd ~'
alias up='cd ..'
alias h='history'
alias grep='grep -E --color=auto'
alias n='nano'

alias noSleep='caffeinate -u -t 3600'



#zsh goodies
setopt auto_cd # cd by typing directory name if it's not a command
setopt correct_all # autocorrect commands
setopt auto_menu # automatically use menu completion
setopt auto_list # automatically list choices on ambiguous completion
setopt always_to_end # move cursor to end if word had one match
zstyle ':completion:*' menu select # select completions with arrow keys
zstyle ':completion:*' group-name '' # group results by category
zstyle ':completion:::::' completer _expand _complete _ignored _approximate # enable approximate matches for completion
#bindkey '^[[3~' delete-char
#bindkey '^[3;5~' delete-char
# load + start compinit
autoload -U compinit && compinit
# automatic cd into directory
setopt autocd
# correct if completion is no possible
zstyle ':completion:*' completer _complete _approximate
zstyle ':bracketed-paste-magic' active-widgets '.self-*
```

**Comment by rwmitchell** [comment]
What is $LANG set to?

**Comment by Tissot11** [comment]
(Response to https://github.com/ohmyzsh/ohmyzsh/discussions/10383#discussioncomment-1580867)

I had to set it up because I couldn't compile codes on a remote cluster. Are these creating problems and how one should proceed for avoiding both problems on the terminal prompt and also compiling code on a remote cluster.

```
export LANG=C
export LC_ALL=C
```

**Reply by rwmitchell** [reply]
I have
```export LANG='en_US.UTF-8'```

**Comment by Tissot11** [comment]
This setting doesn't help me for compiling codes on remote a cluster.

**Reply by mcornella** [reply]
Please reply to someone in the same thread: 
![image](https://user-images.githubusercontent.com/1441704/140053325-be9ab73b-b4b2-458f-b1a5-ab8500851f0c.png)

**Comment by mcornella** [comment, marked as answer]
## To fix the prompt

Don't use `%{%G<emoji%}`, that is going to confuse zsh. I found that it fixed my prompt problems when I removed these and just added the emoji. You should have a proper UTF-8 locale though, [which you don't](https://github.com/ohmyzsh/ohmyzsh/discussions/10383#discussioncomment-1580922).

You also have many `%{` and `%}`, which tell zsh to treat what is inside as not occupying a character, when it's not really necessary (`%F{color}` and `%f` don't need to go inside these). I also removed redundant "new foreground color" / "reset foreground color" sequences without nothing inside (`%F{magenta}%f` is just the same as `%f`).

Lastly, you have two lines setting `PS1`, I've fixed both of them and you choose whichever you prefer:

```diff
-PS1="[%{%F{blue}%}%n%{%f%}@%m-|%F{blue}%w-%F{blue}%*% %F{blue}%f|-%F{magenta}%~%}%{%F{blue}%}]-%F{magneta}%f%{%G🖖%} %f%}%{%G>%}%f "
+PS1="[%F{blue}%n%f@%m-|%F{blue}%w-%F{blue}%*% %f|-%F{magenta}%~%F{blue}]-%f🖖 > "

-PS1="%F{blue}[%}%n%{%f%}@%F{blue}%D{%a %b %e|%X%z(%Z)|}%F{blue}%f%F{magenta}%~%}%{%F{blue}%}]-%F{magneta}%f%{%G🖖%} %f%}%{%G>%}%f "
+PS1="%F{blue}[%n%f@%F{blue}%D{%a %b %e|%X%z(%Z)|}%F{magenta}%~%F{blue}]-%f🖖 > "
```

----

## To fix the locale

As I said, you need a UTF-8 locale. Run `locale -a` to see which locales ending in `.UTF-8` or `utf8` you have, then select one that works for you in terms of language and use that.

If you don't want to have a language, sometimes you have `C.UTF-8` available, which doesn't set a language but it properly uses UTF-8. That is going to depend on what the output of `locale -a`. If you don't have any locales installed that support UTF-8, you'll need to install one.

----

PS. This alias is wrong. To reload zsh, run `exec zsh`, **never `source ~/.zshrc`** which can have side effects:
```zsh
alias reload='source ~/.zshrc'
```

**Comment by Tissot11** [comment]
Thanks for the answer! I switched from `bash` to `zsh` so I was still trying to make things work as they were before. Your suggestions seem to be working. I had forgotten to comment the first PS1. Now, I try your PS1 and PS2 edits and also set the `export LANG='en_US.UTF-8'`. I'll check it if this creates any issues for compiling on a remote cluster. Thanks again!!

**Reply by mcornella** [reply]
This is how you reply to an answer:
![image](https://user-images.githubusercontent.com/1441704/140166121-80e63823-556e-426f-9509-255b8d81c16d.png)

**Reply by Tissot11** [reply]
Both boxes are shown to write a comment. I chose the box that looked bigger to write a reply.
